pkg/volume/csi/csi_plugin.go

```diff
@@ -246,16 +247,90 @@ func (p *csiPlugin) Init(host volume.VolumeHost) error {
 		}
 	}
 
-	// Initializing csiDrivers map and label management channels
+	// Initializing csiDrivers map, label management channels, and migratedDrivers map
 	csiDrivers = csiDriversStore{driversMap: map[string]csiDriver{}}
-	nim = nodeinfomanager.NewNodeInfoManager(host.GetNodeName(), host)
 
-	// TODO(#70514) Init CSINodeInfo object if the CRD exists and create Driver
-	// objects for migrated drivers.
+	var migratedDrivers = map[string](func() bool){
+		csiPlugins.GCEPDDriverName: func() bool {
+			return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) && utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationGCE)
+		},
+		// TODO(leakingtpan): Add AWS migration feature gates and place them here
+	}
+
+	nim = nodeinfomanager.NewNodeInfoManager(host.GetNodeName(), host, migratedDrivers)
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		// This function prevents Kubelet from posting Ready status until CSINodeInfo
+		// is both installed and initialized
+		initializeCSINodeInfo(host)
+	}
 
 	return nil
 }
 
+func initializeCSINodeInfo(host volume.VolumeHost) {
+	kubeClient := host.GetKubeClient()
+	if kubeClient == nil || len(host.GetPodsDir()) == 0 {
+		// If we're NOT on the Kubelet OR KubeClient DOESNT exist
+		klog.V(4).Infof("Skipping CSINodeInfo initialization, not running on Kubelet or KubeClient doesn't exist")
+		return
+	}
+
+	go func() {
+		host.SetKubeletError(fmt.Errorf("Could not find CSINodeInfo CRD. Please install the CSINodeInfo CRD"))
```

Comment:
setting the error inside the gofunc is racy... need to initialize it synchronously before starting the gofunc (and start it at something more accurate like "CSINodeInfo availability not yet verified"). don't include "CRD" in the error message

Reply:
I tried doing this earlier, however the issue then is then since clearing the error is the job of the CSI Volume Plugin, if the user is to turn CSI off, the error may not be cleared...

I guess it could clear it inside the init. And also there should be some logic outside of volume plugin initialization that clears it if CSI is off in general. How does that sound?

Reply:
the only way to turn it off is via the feature gate, right? doesn't that requires a restart of the kubelet, which would mean initializeCSINodeInfo() would not get called on startup, and the error would remain nil as it is on startup